core/include/kernel/interrupt.h

```diff
@@ -164,4 +181,117 @@ static inline struct itr_handler *itr_alloc_add(size_t it,
 				       0);
 }
 
+/*
+ * Interrupt controller chip API functions
+ */
+
+/*
+ * interrupt_call_handlers() - Call registered handlers for an interrupt
+ * @chip	Interrupt controller
+ * @itr_num	Interrupt number
+ *
+ * This function is called from an interrupt context by a primary interrupt
+ * handler. This function calls the handlers registered for that interrupt
+ */
+void interrupt_call_handlers(struct itr_chip *chip, size_t itr_num);
+
+/*
+ * interrupt_mask() - Mask a interrupt
+ * @chip	Interrupt controller
+ * @itr_num	Interrupt number
+ * This function may be called in interrupt context
+ */
+static inline void interrupt_mask(struct itr_chip *chip, size_t itr_num)
+{
+	if (chip->ops->mask)
```

Comment:
Why not make this function pointer mandatory? Same for the other ops below.

Reply:
right, let's make it mandatory

Reply:
@jenswi-linaro: Should we really mandate mask/unmask operators? GIC, Atmel SAIC and Hafnium HFIC don't provide specific interrupt masking means so I need to use their relate interrupt disable/enable function for these operators. Does it make sence for these main interrupt controllers?

Reply:
I guess my confusion regarding the mask/unmask operators stems from this, I don't see what it's supposed to do.

Reply:
Some interrupt controllers are able to mask/unmask interrupt, so it's fair to provide a handler for that.

For consistency, I think we should mandate `mask`/`unmask` operators and change `interrupt_call_handlers()` to mask unhandled interrupt, not disable them, as `.disable` operator has only mask.unmask are expected to be called from uninterruptible contexts.

Regarding GIC, SAIC and other main controller driver, they can't mask individual interrupts, and no one ask them to do so, so let use their disable/enable function handlers as mask/unmask handlers, as best effort. It think it is fair this way.

Reply:
> Some interrupt controllers are able to mask/unmask interrupt, so it's fair to provide a handler for that.

Out of curiosity, could you name one?

Reply:
Actually, I must admit I did not find any. I was confused by documentation that sometimes uses mask and enable terminology for the same configuration means. Do you think we can remove mask/unmask and keep only the enable/disable operators? Until someone request we splits these couples.

Reply:
Thanks for your effort. Yes, I think we can remove mask/unmask for now.

Reply:
@jenswi-linaro @etienne-lms
I have a real use case that requires mask/unmask to be split from enable/disable.

An I2C based consumer sends itr's through GPIO to the itr hierarchy made by:
- edge triggered EXTI itr controller, that in turn sends its itr to
- level triggered ARM GIC.

To ack the itr in the consumer through I2C I have to rely on a bottom half using async notif.

I cannot "disable" the itr in the EXTI otherwise (due to edge base trigger) I will loose a new incoming itr
between the consumer's ack and the EXTI "enable".
So I have to rely on the "disable" of GIC, while keeping EXTI enabled.

The top/bottom half sequence is:
- top half:
  .1) "mask" itr, that means "ack the edge" on EXTI and "disable" the itr on GIC
  .2) notify bottom half
- bottom half:
  .3) ack the consumer's itr through I2C
  .4) "unmask" itr, that "does nothing" on EXTI and "enables" GIC

A new edge itr occurring between 3) and 4) is recorded by EXTI but propagated by GIC only after 4).

Level triggered itr controllers don't need to differentiate enable/disable from mask/unmask
because they will not loose an incoming level itr; if the level is still there, it will trigger the itr
once controller re-enables.
Edge triggered itr controllers need mask to block the propagation of new itr and ack internally the old
itr, while still sensing for new edges. Unmask will eventually propagate the newly recorded itr.

In my hierarchy above, EXTI does not have a mask feature, but relies on GIC for this.

GIC should register mask/unmask that corresponds to current gic_op_enable/gic_op_disable, as
already done in patch "driver: gic: implement mask/unmask handler".
EXTI will register enable/disable that operates as real enable/disable of EXTI and also
register an empty "unmask" and a "mask" that ack the itr.

Reply:
Thank @borneoa for the details.
So I think it is better to keep the 4 mask/unmaks/enable/disable handlers and relates API function `interrupt_xxx()`. An interrupt controller that has registered to handler to the GIC can offer real interrupt mask/unmask/enable/disable services to its own consumers.

I have created P-R https://github.com/OP-TEE/optee_os/pull/6096 to  specifically review these API functions.